src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
-              // if we can remove the task (aka it has not been removed already), we can execute it
-              nextExecuteTask.executing();
-              return nextExecuteTask;
+            } else {
+              LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextTaskDelay);
             }
-          } else if (executeQueue.remove(nextExecuteTask)) {
-            // if we can remove the task (aka it has not been removed already), we can execute it
-            nextExecuteTask.executing();
-            return nextExecuteTask;
           }
-        } else if (nextScheduledTask != null) {
-          if (nextScheduledTask.getDelay(TimeUnit.MILLISECONDS) <= 0) {
-            synchronized (scheduleQueue.getModificationLock()) {
-              // scheduled tasks must be removed, and call .executing() while holding the lock
-              if (scheduleQueue.remove(nextScheduledTask)) {
-                nextScheduledTask.executing();
-                return nextScheduledTask;
-              }
+        } else if (nextHighPriorityTask == null) {
+          // only low priority task is queued, so run if ready
+          long nextTaskDelay = nextLowPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+          if (nextTaskDelay <= 0) {
+            if (nextLowPriorityTask.canExecute()) {
+              return nextLowPriorityTask;
             }
           } else {
-            workerPool.handleEstimatedTimeTillNextTask(this, nextScheduledTask.getDelayEstimateInMs());
-            LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextScheduledTask.getDelay(TimeUnit.MILLISECONDS));
+            LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextTaskDelay);
           }
         } else {
-          workerPool.handleEstimatedTimeTillNextTask(this, Long.MAX_VALUE);
-          LockSupport.park();
+          // both tasks are available, so see which one makes sense to run
+          long nextHighDelay;
+          long nextLowDelay;
+          ClockWrapper.stopForcingUpdate();
+          try {
+            nextHighDelay = nextHighPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+            nextLowDelay = nextLowPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+          } finally {
+            ClockWrapper.resumeForcingUpdate();
+          }
+          if (nextHighDelay <= 0) {
+            // high task is ready, make sure it should be picked
+            if (nextHighDelay <= nextLowDelay || nextLowDelay > maxWaitForLowPriorityInMs * -1) {
+              if (nextHighPriorityTask.canExecute()) {
+                return nextHighPriorityTask;
+              }
+            } else if (nextLowPriorityTask.canExecute()) {
+              return nextLowPriorityTask;
+            }
+          } else if (nextLowDelay <= 0 && nextLowPriorityTask.canExecute()) {
+            return nextLowPriorityTask;
+          } else {
+            if (nextLowDelay < nextHighDelay) {
+              LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextLowDelay);
```

Comment:
seems like we could hint at what we expect to be ran in the next loop here.  possibly bypassing a lot of logic for the first loop.  Not sure exactly how it would work but could be worth looking into.